src/gpt.py

```diff
@@ -0,0 +1,312 @@
+# Initialize globals
+def initialize(client):
+    global gpt_lock
+    global context_data
+    global encoding
+    global self_user
+    global banned_word_pattern
+    global word_trigger_pattern
+    global user_trigger_pattern
+    global user_info_map
+
+    openai.api_key = config.openai_api_key
+    # Global lock for message response
+    gpt_lock = asyncio.Lock()
+    # Load manual hidden data
+    try:
+        with open(config.gpt_context_file) as f:
+            context_data = json.load(f)
+    except (FileNotFoundError, json.decoder.JSONDecodeError) as e:
+        print(
+            f"ERROR: Issue loading {config.gpt_context_file}. GPT capabilities will be disabled.\n{e}"
+        )
+        # Signal context data was not loaded correctly
+        context_data = None
+        return
+    # Preload tokenizer
+    encoding = tiktoken.encoding_for_model(config.OPENAI_MODEL)
+    # Store bot user
+    self_user = client.user
+    # Cache regex for banned words
+    banned_word_pattern = re.compile(
+        r"\b{}\b".format("|".join(context_data["banned_phrases"]))
+    )
+
+    # Cache regex for word triggers
+    word_trigger_pattern = re.compile(
+        r"\b({})\b".format("|".join(context_data["word_triggers"].keys()))
+    )
+
+    # Cache regex for user info triggers
+    user_trigger_pattern = re.compile(
+        r"\b({})\b".format(
+            "|".join(
+                [
+                    user["name"].lower()
+                    for user in context_data["user_info"].values()
+                    if "info" in user
+                ]
+            )
+        )
+    )
+    # Store map for user info triggers
+    user_info_map = {
+        user["name"].lower(): user["info"].lower()
+        for user in context_data["user_info"].values()
+        if ("info" in user and "name" in user)
+    }
+
+
+# Check if a message should not be allow in API history
+# Currently this is just if it contains banned phrases
+def disallowed_message(message: Message) -> bool:
+    return banned_word_pattern.search(message.content.lower()) is not None
+
+
+# Get the name we should call the user
+def get_name(user):
+    user_info = context_data["user_info"]
+    id = str(user.id)
+    if id in user_info and "name" in user_info[id]:
+        return user_info[id]["name"]
+    return user.name
+
+
+# Get totally random triggers about how to act:
+def get_random_context():
+    context = []
+    value = random.random()
+    if value < 0.25:
+        context.append("Respond in 10 words or less")
+    elif value < 0.5:
+        context.append("Be succinct")
+    elif value < 0.75:
+        context.append("Don't be too wordy")
+    else:
+        context.append("Write long detailed responses")
+    return context
+
+
+# Get context about the server
+async def get_server_context(message: Message) -> str:
+    result = []
+    # Load server event info
+    if message.guild and message.guild.scheduled_events:
+        # No idea if these are sorted by time
```

Comment:
I think my preferred solution would be to always take the soonest event, then if there's no ":" splitting character we say

"next event" instead of "next bust event" or something

Reply:
Oh, to give the bot context on any event? Indeed, that could be interesting!

Reply:
Added